From 25304ebd8bdd6230ed8aba8c93ee4a6b5ef1c665 Mon Sep 17 00:00:00 2001
From: James Chapman <jachapma@redhat.com>
Date: Fri, 30 May 2025 11:12:43 +0000
Subject: [PATCH] Issue 6641 - modrdn fails when a user is member of multiple
 groups (#6643)

Bug description:
Rename of a user that is member of multiple AM groups fail when MO and
RI plugins are enabled.

Fix description:
MO plugin - After updating the entry member attribute, check the return
value. Retry the delete if the attr value exists and retry the add if the
attr value is missing.

RI plugin - A previous commit checked if the attr value was not present
before adding a mod. This commit was reverted in favour of overriding
the internal op return value, consistent with other plugins.

CI test from Viktor Ashirov <vashirov@redhat.com>

Fixes: https://github.com/389ds/389-ds-base/issues/6641
Relates: https://github.com/389ds/389-ds-base/issues/6566

Reviewed by: @progier389, @tbordaz, @vashirov (Thank you)
---
 .../tests/suites/plugins/modrdn_test.py       | 174 ++++++++++++++++++
 ldap/servers/plugins/automember/automember.c  |  11 +-
 ldap/servers/plugins/memberof/memberof.c      | 124 +++++--------
 ldap/servers/plugins/referint/referint.c      |  43 +++--
 ldap/servers/slapd/modify.c                   |  51 +++++
 ldap/servers/slapd/slapi-plugin.h             |   1 +
 6 files changed, 303 insertions(+), 101 deletions(-)
 create mode 100644 dirsrvtests/tests/suites/plugins/modrdn_test.py

diff --git a/dirsrvtests/tests/suites/plugins/modrdn_test.py b/dirsrvtests/tests/suites/plugins/modrdn_test.py
new file mode 100644
index 000000000..be79b0c3c
--- /dev/null
+++ b/dirsrvtests/tests/suites/plugins/modrdn_test.py
@@ -0,0 +1,174 @@
+# --- BEGIN COPYRIGHT BLOCK ---
+# Copyright (C) 2025 Red Hat, Inc.
+# All rights reserved.
+#
+# License: GPL (version 3 or any later version).
+# See LICENSE for details.
+# --- END COPYRIGHT BLOCK ---
+#
+import pytest
+from lib389.topologies import topology_st
+from lib389._constants import DEFAULT_SUFFIX
+from lib389.idm.group import Groups
+from lib389.idm.user import nsUserAccounts
+from lib389.plugins import (
+    AutoMembershipDefinitions,
+    AutoMembershipPlugin,
+    AutoMembershipRegexRules,
+    MemberOfPlugin,
+    ReferentialIntegrityPlugin,
+)
+
+pytestmark = pytest.mark.tier1
+
+USER_PROPERTIES = {
+    "uid": "userwith",
+    "cn": "userwith",
+    "uidNumber": "1000",
+    "gidNumber": "2000",
+    "homeDirectory": "/home/testuser",
+    "displayName": "test user",
+}
+
+
+def test_modrdn_of_a_member_of_2_automember_groups(topology_st):
+    """Test that a member of 2 automember groups can be renamed
+
+    :id: 0e40bdc4-a2d2-4bb8-8368-e02c8920bad2
+
+    :setup: Standalone instance
+
+    :steps:
+        1. Enable automember plugin
+        2. Create definiton for users with A in the name
+        3. Create regex rule for users with A in the name
+        4. Create definiton for users with Z in the name
+        5. Create regex rule for users with Z in the name
+        6. Enable memberof plugin
+        7. Enable referential integrity plugin
+        8. Restart the instance
+        9. Create groups
+        10. Create users userwitha, userwithz, userwithaz
+        11. Rename userwithaz
+
+    :expectedresults:
+        1. Success
+        2. Success
+        3. Success
+        4. Success
+        5. Success
+        6. Success
+        7. Success
+        8. Success
+        9. Success
+        10. Success
+        11. Success
+    """
+    inst = topology_st.standalone
+
+    # Enable automember plugin
+    automember_plugin = AutoMembershipPlugin(inst)
+    automember_plugin.enable()
+
+    # Create definiton for users with A in the name
+    automembers = AutoMembershipDefinitions(inst)
+    automember = automembers.create(
+        properties={
+            "cn": "userswithA",
+            "autoMemberScope": DEFAULT_SUFFIX,
+            "autoMemberFilter": "objectclass=posixAccount",
+            "autoMemberGroupingAttr": "member:dn",
+        }
+    )
+
+    # Create regex rule for users with A in the name
+    automembers_regex_rule = AutoMembershipRegexRules(inst, f"{automember.dn}")
+    automembers_regex_rule.create(
+        properties={
+            "cn": "userswithA",
+            "autoMemberInclusiveRegex": ["cn=.*a.*"],
+            "autoMemberTargetGroup": [f"cn=userswithA,ou=Groups,{DEFAULT_SUFFIX}"],
+        }
+    )
+
+    # Create definiton for users with Z in the name
+    automember = automembers.create(
+        properties={
+            "cn": "userswithZ",
+            "autoMemberScope": DEFAULT_SUFFIX,
+            "autoMemberFilter": "objectclass=posixAccount",
+            "autoMemberGroupingAttr": "member:dn",
+        }
+    )
+
+    # Create regex rule for users with Z in the name
+    automembers_regex_rule = AutoMembershipRegexRules(inst, f"{automember.dn}")
+    automembers_regex_rule.create(
+        properties={
+            "cn": "userswithZ",
+            "autoMemberInclusiveRegex": ["cn=.*z.*"],
+            "autoMemberTargetGroup": [f"cn=userswithZ,ou=Groups,{DEFAULT_SUFFIX}"],
+        }
+    )
+
+    # Enable memberof plugin
+    memberof_plugin = MemberOfPlugin(inst)
+    memberof_plugin.enable()
+
+    # Enable referential integrity plugin
+    referint_plugin = ReferentialIntegrityPlugin(inst)
+    referint_plugin.enable()
+
+    # Restart the instance
+    inst.restart()
+
+    # Create groups
+    groups = Groups(inst, DEFAULT_SUFFIX)
+    groupA = groups.create(properties={"cn": "userswithA"})
+    groupZ = groups.create(properties={"cn": "userswithZ"})
+
+    # Create users
+    users = nsUserAccounts(inst, DEFAULT_SUFFIX)
+
+    # userwitha
+    user_props = USER_PROPERTIES.copy()
+    user_props.update(
+        {
+            "uid": USER_PROPERTIES["uid"] + "a",
+            "cn": USER_PROPERTIES["cn"] + "a",
+        }
+    )
+    user = users.create(properties=user_props)
+
+    # userwithz
+    user_props.update(
+        {
+            "uid": USER_PROPERTIES["uid"] + "z",
+            "cn": USER_PROPERTIES["cn"] + "z",
+        }
+    )
+    user = users.create(properties=user_props)
+
+    # userwithaz
+    user_props.update(
+        {
+            "uid": USER_PROPERTIES["uid"] + "az",
+            "cn": USER_PROPERTIES["cn"] + "az",
+        }
+    )
+    user = users.create(properties=user_props)
+    user_orig_dn = user.dn
+
+    # Rename userwithaz
+    user.rename(new_rdn="uid=userwith")
+    user_new_dn = user.dn
+
+    assert user.get_attr_val_utf8("uid") != "userwithaz"
+
+    # Check groups contain renamed username
+    assert groupA.is_member(user_new_dn)
+    assert groupZ.is_member(user_new_dn)
+
+    # Check groups dont contain original username
+    assert not groupA.is_member(user_orig_dn)
+    assert not groupZ.is_member(user_orig_dn)
diff --git a/ldap/servers/plugins/automember/automember.c b/ldap/servers/plugins/automember/automember.c
index 38f817e5d..f900db7f2 100644
--- a/ldap/servers/plugins/automember/automember.c
+++ b/ldap/servers/plugins/automember/automember.c
@@ -1755,13 +1755,12 @@ automember_update_member_value(Slapi_Entry *member_e, const char *group_dn, char
         }
 
         mod_pb = slapi_pblock_new();
-        slapi_modify_internal_set_pb(mod_pb, group_dn,
-                                     mods, 0, 0, automember_get_plugin_id(), 0);
-        slapi_modify_internal_pb(mod_pb);
-        slapi_pblock_get(mod_pb, SLAPI_PLUGIN_INTOP_RESULT, &result);
+        /* Do a single mod with error overrides for DEL/ADD */
+        result = slapi_single_modify_internal_override(mod_pb, slapi_sdn_new_dn_byval(group_dn), mods,
+                                                        automember_get_plugin_id(), 0);
 
         if(add){
-            if ((result != LDAP_SUCCESS) && (result != LDAP_TYPE_OR_VALUE_EXISTS)) {
+            if (result != LDAP_SUCCESS) {
                 slapi_log_err(SLAPI_LOG_ERR, AUTOMEMBER_PLUGIN_SUBSYSTEM,
                               "automember_update_member_value - Unable to add \"%s\" as "
                               "a \"%s\" value to group \"%s\" (%s).\n",
@@ -1771,7 +1770,7 @@ automember_update_member_value(Slapi_Entry *member_e, const char *group_dn, char
             }
         } else {
             /* delete value */
-            if ((result != LDAP_SUCCESS) && (result != LDAP_NO_SUCH_ATTRIBUTE)) {
+            if (result != LDAP_SUCCESS) {
                 slapi_log_err(SLAPI_LOG_ERR, AUTOMEMBER_PLUGIN_SUBSYSTEM,
                               "automember_update_member_value - Unable to delete \"%s\" as "
                               "a \"%s\" value from group \"%s\" (%s).\n",
diff --git a/ldap/servers/plugins/memberof/memberof.c b/ldap/servers/plugins/memberof/memberof.c
index 35fd1a4a0..16dae2195 100644
--- a/ldap/servers/plugins/memberof/memberof.c
+++ b/ldap/servers/plugins/memberof/memberof.c
@@ -964,7 +964,6 @@ modify_need_fixup(int set)
         mod_pb, memberof_get_config_area(),
         mods, 0, 0,
         memberof_get_plugin_id(), SLAPI_OP_FLAG_FIXUP|SLAPI_OP_FLAG_BYPASS_REFERRALS);
-
     slapi_modify_internal_pb(mod_pb);
     slapi_pblock_get(mod_pb, SLAPI_PLUGIN_INTOP_RESULT, &rc);
     slapi_pblock_destroy(mod_pb);
@@ -1492,18 +1491,9 @@ memberof_del_dn_type_callback(Slapi_Entry *e, void *callback_data)
     mod.mod_op = LDAP_MOD_DELETE;
     mod.mod_type = ((memberof_del_dn_data *)callback_data)->type;
     mod.mod_values = val;
-
-    slapi_modify_internal_set_pb_ext(
-        mod_pb, slapi_entry_get_sdn(e),
-        mods, 0, 0,
-        memberof_get_plugin_id(), SLAPI_OP_FLAG_BYPASS_REFERRALS);
-
-    slapi_modify_internal_pb(mod_pb);
-
-    slapi_pblock_get(mod_pb,
-                     SLAPI_PLUGIN_INTOP_RESULT,
-                     &rc);
-
+    /* Internal mod with error overrides for DEL/ADD */
+    rc = slapi_single_modify_internal_override(mod_pb, slapi_entry_get_sdn(e), mods,
+                                                memberof_get_plugin_id(), SLAPI_OP_FLAG_BYPASS_REFERRALS);
     slapi_pblock_destroy(mod_pb);
 
     if (rc == LDAP_NO_SUCH_ATTRIBUTE && val[0] == NULL) {
@@ -1976,6 +1966,7 @@ memberof_replace_dn_type_callback(Slapi_Entry *e, void *callback_data)
 
     return rc;
 }
+
 LDAPMod **
 my_copy_mods(LDAPMod **orig_mods)
 {
@@ -2784,33 +2775,6 @@ memberof_modop_one_replace_r(Slapi_PBlock *pb, MemberOfConfig *config, int mod_o
                 replace_mod.mod_values = replace_val;
             }
             rc = memberof_add_memberof_attr(mods, op_to, config->auto_add_oc);
-            if (rc == LDAP_NO_SUCH_ATTRIBUTE || rc == LDAP_TYPE_OR_VALUE_EXISTS) {
-                if (rc == LDAP_TYPE_OR_VALUE_EXISTS) {
-                    /*
-                     * For some reason the new modrdn value is present, so retry
-                     * the delete by itself and ignore the add op by tweaking
-                     * the mod array.
-                     */
-                    mods[1] = NULL;
-                    rc = memberof_add_memberof_attr(mods, op_to, config->auto_add_oc);
-                } else {
-                    /*
-                     * The memberof value to be replaced does not exist so just
-                     * add the new value.  Shuffle the mod array to apply only
-                     * the add operation.
-                     */
-                    mods[0] = mods[1];
-                    mods[1] = NULL;
-                    rc = memberof_add_memberof_attr(mods, op_to, config->auto_add_oc);
-                    if (rc == LDAP_TYPE_OR_VALUE_EXISTS) {
-                        /*
-                         * The entry already has the expected memberOf value, no
-                         * problem just return success.
-                         */
-                        rc = LDAP_SUCCESS;
-                    }
-                }
-            }
         }
     }
 
@@ -4461,43 +4425,57 @@ memberof_add_memberof_attr(LDAPMod **mods, const char *dn, char *add_oc)
     Slapi_PBlock *mod_pb = NULL;
     int added_oc = 0;
     int rc = 0;
+    LDAPMod *single_mod[2];
 
-    while (1) {
-        mod_pb = slapi_pblock_new();
-        slapi_modify_internal_set_pb(
-            mod_pb, dn, mods, 0, 0,
-            memberof_get_plugin_id(), SLAPI_OP_FLAG_BYPASS_REFERRALS);
-        slapi_modify_internal_pb(mod_pb);
-
-        slapi_pblock_get(mod_pb, SLAPI_PLUGIN_INTOP_RESULT, &rc);
-        if (rc == LDAP_OBJECT_CLASS_VIOLATION) {
-            if (!add_oc || added_oc) {
-                /*
-                 * We aren't auto adding an objectclass, or we already
-                 * added the objectclass, and we are still failing.
-                 */
+    if (!dn || !mods) {
+        slapi_log_err(SLAPI_LOG_ERR, MEMBEROF_PLUGIN_SUBSYSTEM,
+                    "Invalid argument: %s%s is NULL\n",
+                    !dn ? "dn " : "",
+                    !mods ? "mods " : "");
+        return LDAP_PARAM_ERROR;
+    }
+
+
+    mod_pb = slapi_pblock_new();
+    /* Split multiple mods into individual mod operations */
+    for (size_t i = 0; (mods != NULL) && (mods[i] != NULL); i++) {
+        single_mod[0] = mods[i];
+        single_mod[1] = NULL;
+
+        while (1) {
+            slapi_pblock_init(mod_pb);
+
+            /* Internal mod with error overrides for DEL/ADD */
+            rc = slapi_single_modify_internal_override(mod_pb, slapi_sdn_new_normdn_byref(dn), single_mod,
+                                                        memberof_get_plugin_id(), SLAPI_OP_FLAG_BYPASS_REFERRALS);
+            if (rc == LDAP_OBJECT_CLASS_VIOLATION) {
+                if (!add_oc || added_oc) {
+                    /*
+                    * We aren't auto adding an objectclass, or we already
+                    * added the objectclass, and we are still failing.
+                    */
+                    break;
+                }
+                rc = memberof_add_objectclass(add_oc, dn);
+                slapi_log_err(SLAPI_LOG_WARNING, MEMBEROF_PLUGIN_SUBSYSTEM,
+                        "Entry %s - schema violation caught - repair operation %s\n",
+                        dn ? dn : "unknown",
+                        rc ? "failed" : "succeeded");
+                if (rc) {
+                    /* Failed to add objectclass */
+                    rc = LDAP_OBJECT_CLASS_VIOLATION;
+                    break;
+                }
+                added_oc = 1;
+            } else if (rc) {
+                /* Some other fatal error */
+                slapi_log_err(SLAPI_LOG_PLUGIN, MEMBEROF_PLUGIN_SUBSYSTEM,
+                            "memberof_add_memberof_attr - Internal modify failed. rc=%d\n", rc);
                 break;
-            }
-            rc = memberof_add_objectclass(add_oc, dn);
-            slapi_log_err(SLAPI_LOG_WARNING, MEMBEROF_PLUGIN_SUBSYSTEM,
-                    "Entry %s - schema violation caught - repair operation %s\n",
-                    dn ? dn : "unknown",
-                    rc ? "failed" : "succeeded");
-            if (rc) {
-                /* Failed to add objectclass */
-                rc = LDAP_OBJECT_CLASS_VIOLATION;
+            } else {
+                /* success */
                 break;
             }
-            added_oc = 1;
-            slapi_pblock_destroy(mod_pb);
-        } else if (rc) {
-            /* Some other fatal error */
-            slapi_log_err(SLAPI_LOG_PLUGIN, MEMBEROF_PLUGIN_SUBSYSTEM,
-                          "memberof_add_memberof_attr - Internal modify failed. rc=%d\n", rc);
-            break;
-        } else {
-            /* success */
-            break;
         }
     }
     slapi_pblock_destroy(mod_pb);
diff --git a/ldap/servers/plugins/referint/referint.c b/ldap/servers/plugins/referint/referint.c
index 218863ea5..a2f2e4706 100644
--- a/ldap/servers/plugins/referint/referint.c
+++ b/ldap/servers/plugins/referint/referint.c
@@ -712,19 +712,28 @@ static int
 _do_modify(Slapi_PBlock *mod_pb, Slapi_DN *entrySDN, LDAPMod **mods)
 {
     int rc = 0;
+    LDAPMod *mod[2];
 
-    slapi_pblock_init(mod_pb);
+    /* Split multiple modifications into individual modify operations */
+    for (size_t i = 0; (mods != NULL) && (mods[i] != NULL); i++) {
+        mod[0] = mods[i];
+        mod[1] = NULL;
 
-    if (allow_repl) {
-        /* Must set as a replicated operation */
-        slapi_modify_internal_set_pb_ext(mod_pb, entrySDN, mods, NULL, NULL,
-                                         referint_plugin_identity, OP_FLAG_REPLICATED);
-    } else {
-        slapi_modify_internal_set_pb_ext(mod_pb, entrySDN, mods, NULL, NULL,
-                                         referint_plugin_identity, 0);
+        slapi_pblock_init(mod_pb);
+
+        /* Do a single mod with error overrides for DEL/ADD */
+        if (allow_repl) {
+            rc = slapi_single_modify_internal_override(mod_pb, entrySDN, mod,
+                                                        referint_plugin_identity, OP_FLAG_REPLICATED);
+        } else {
+            rc = slapi_single_modify_internal_override(mod_pb, entrySDN, mod,
+                                                        referint_plugin_identity, 0);
+        }
+
+        if (rc != LDAP_SUCCESS) {
+            return rc;
+        }
     }
-    slapi_modify_internal_pb(mod_pb);
-    slapi_pblock_get(mod_pb, SLAPI_PLUGIN_INTOP_RESULT, &rc);
 
     return rc;
 }
@@ -924,7 +933,6 @@ _update_all_per_mod(Slapi_DN *entrySDN, /* DN of the searched entry */
 {
     Slapi_Mods *smods = NULL;
     char *newDN = NULL;
-    struct berval bv = {0};
     char **dnParts = NULL;
     char *sval = NULL;
     char *newvalue = NULL;
@@ -1027,30 +1035,21 @@ _update_all_per_mod(Slapi_DN *entrySDN, /* DN of the searched entry */
             }
             /* else: normalize_rc < 0) Ignore the DN normalization error for now. */
 
-            bv.bv_val = newDN;
-            bv.bv_len = strlen(newDN);
             p = PL_strstr(sval, slapi_sdn_get_ndn(origDN));
             if (p == sval) {
                 /* (case 1) */
                 slapi_mods_add_string(smods, LDAP_MOD_DELETE, attrName, sval);
-                /* Add only if the attr value does not exist */
-                if (VALUE_PRESENT != attr_value_find_wsi(attr, &bv, &v)) {
-                    slapi_mods_add_string(smods, LDAP_MOD_ADD, attrName, newDN);
-                }
+                slapi_mods_add_string(smods, LDAP_MOD_ADD, attrName, newDN);
             } else if (p) {
                 /* (case 2) */
                 slapi_mods_add_string(smods, LDAP_MOD_DELETE, attrName, sval);
                 *p = '\0';
                 newvalue = slapi_ch_smprintf("%s%s", sval, newDN);
-                /* Add only if the attr value does not exist */
-                if (VALUE_PRESENT != attr_value_find_wsi(attr, &bv, &v)) {
-                    slapi_mods_add_string(smods, LDAP_MOD_ADD, attrName, newvalue);
-                }
+                slapi_mods_add_string(smods, LDAP_MOD_ADD, attrName, newvalue);
                 slapi_ch_free_string(&newvalue);
             }
             /* else: value does not include the modified DN.  Ignore it. */
             slapi_ch_free_string(&sval);
-            bv = (struct berval){0};
         }
         rc = _do_modify(mod_pb, entrySDN, slapi_mods_get_ldapmods_byref(smods));
         if (rc) {
diff --git a/ldap/servers/slapd/modify.c b/ldap/servers/slapd/modify.c
index 55131d77c..8241fce28 100644
--- a/ldap/servers/slapd/modify.c
+++ b/ldap/servers/slapd/modify.c
@@ -489,6 +489,57 @@ slapi_modify_internal_set_pb_ext(Slapi_PBlock *pb, const Slapi_DN *sdn, LDAPMod
     slapi_pblock_set(pb, SLAPI_PLUGIN_IDENTITY, plugin_identity);
 }
 
+/* Performs a single LDAP modify operation with error overrides.
+ *
+ * If specific errors occur, such as attempting to add an existing attribute or
+ * delete a non-existent one, the function overrides the error and returns success:
+ *   - LDAP_MOD_ADD -> LDAP_TYPE_OR_VALUE_EXISTS (ignored)
+ *   - LDAP_MOD_DELETE -> LDAP_NO_SUCH_ATTRIBUTE (ignored)
+ *
+ * Any other errors encountered during the operation will be returned as-is.
+ */
+int 
+slapi_single_modify_internal_override(Slapi_PBlock *pb, const Slapi_DN *sdn, LDAPMod **mod, Slapi_ComponentId *plugin_id, int op_flags)
+{
+    int rc = 0;
+    int result = 0;
+    int result_reset = 0;
+    int mod_op = 0;
+
+    if (!pb || !sdn || !mod || !mod[0]) {
+        slapi_log_err(SLAPI_LOG_ERR, "slapi_single_modify_internal_override",
+                    "Invalid argument: %s%s%s%s is NULL\n",
+                    !pb ? "pb " : "",
+                    !sdn ? "sdn " : "",
+                    !mod ? "mod " : "",
+                    !mod[0] ? "mod[0] " : "");
+
+        return LDAP_PARAM_ERROR;
+    }
+
+    slapi_modify_internal_set_pb_ext(pb, sdn, mod, NULL, NULL, plugin_id, op_flags);
+    slapi_modify_internal_pb(pb);
+    slapi_pblock_get(pb, SLAPI_PLUGIN_INTOP_RESULT, &result);
+
+    if (result != LDAP_SUCCESS) {
+        mod_op = mod[0]->mod_op & LDAP_MOD_OP;
+        if ((mod_op == LDAP_MOD_ADD && result == LDAP_TYPE_OR_VALUE_EXISTS) ||
+            (mod_op == LDAP_MOD_DELETE && result == LDAP_NO_SUCH_ATTRIBUTE)) {
+                slapi_log_err(SLAPI_LOG_PLUGIN, "slapi_single_modify_internal_override",
+                            "Overriding return code - plugin:%s dn:%s mod_op:%d result:%d\n",
+                            plugin_id ? plugin_id->sci_component_name : "unknown",
+                            sdn ? sdn->udn : "unknown", mod_op, result);
+
+                slapi_pblock_set(pb, SLAPI_PLUGIN_INTOP_RESULT, &result_reset);
+                rc = LDAP_SUCCESS;
+            } else {
+                rc = result;
+            }
+    }
+
+    return rc;
+}
+
 /* Helper functions */
 
 static int
diff --git a/ldap/servers/slapd/slapi-plugin.h b/ldap/servers/slapd/slapi-plugin.h
index de461706c..c30a2b8ec 100644
--- a/ldap/servers/slapd/slapi-plugin.h
+++ b/ldap/servers/slapd/slapi-plugin.h
@@ -5955,6 +5955,7 @@ void slapi_add_entry_internal_set_pb(Slapi_PBlock *pb, Slapi_Entry *e, LDAPContr
 int slapi_add_internal_set_pb(Slapi_PBlock *pb, const char *dn, LDAPMod **attrs, LDAPControl **controls, Slapi_ComponentId *plugin_identity, int operation_flags);
 void slapi_modify_internal_set_pb(Slapi_PBlock *pb, const char *dn, LDAPMod **mods, LDAPControl **controls, const char *uniqueid, Slapi_ComponentId *plugin_identity, int operation_flags);
 void slapi_modify_internal_set_pb_ext(Slapi_PBlock *pb, const Slapi_DN *sdn, LDAPMod **mods, LDAPControl **controls, const char *uniqueid, Slapi_ComponentId *plugin_identity, int operation_flags);
+int slapi_single_modify_internal_override(Slapi_PBlock *pb, const Slapi_DN *sdn, LDAPMod **mod, Slapi_ComponentId *plugin_identity, int operation_flags);
 /**
  * Set \c Slapi_PBlock to perform modrdn/rename internally
  *
-- 
2.49.0

